ports/atmel-samd/common-hal/watchdog/WatchDogTimer.c

```diff
@@ -0,0 +1,312 @@
+    ;                                 // Wait for synchronization
+  USB->DEVICE.CTRLA.bit.RUNSTDBY = 0; // Deactivate run on standby
+  USB->DEVICE.CTRLA.bit.ENABLE = 1;   // Enable the USB peripheral
+  while (USB->DEVICE.SYNCBUSY.bit.ENABLE)
+    ; // Wait for synchronization
+#else
+  // Generic clock generator 2, divisor = 32 (2^(DIV+1))
+  GCLK->GENDIV.reg = GCLK_GENDIV_ID(2) | GCLK_GENDIV_DIV(4);
+  // Enable clock generator 2 using low-power 32KHz oscillator.
+  // With /32 divisor above, this yields 1024Hz(ish) clock.
+  GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(2) | GCLK_GENCTRL_GENEN |
+                      GCLK_GENCTRL_SRC_OSCULP32K | GCLK_GENCTRL_DIVSEL;
+  while (GCLK->STATUS.bit.SYNCBUSY)
+    ;
+  // WDT clock = clock gen 2
+  GCLK->CLKCTRL.reg =
+      GCLK_CLKCTRL_ID_WDT | GCLK_CLKCTRL_CLKEN | GCLK_CLKCTRL_GEN_GCLK2;
+
+  // Enable WDT early-warning interrupt
+  NVIC_DisableIRQ(WDT_IRQn);
+  NVIC_ClearPendingIRQ(WDT_IRQn);
+  NVIC_SetPriority(WDT_IRQn, 0); // Top priority
+  NVIC_EnableIRQ(WDT_IRQn);
+#endif
+
+  _watchdog_samd_initialized = true;
+}
+
+STATIC int _watchdog_samd_enable(int maxPeriodMS, bool earlyWarning) {
+  // Enable the watchdog with a period up to the specified max period in
+  // milliseconds.
+
+  // Review the watchdog section from the SAMD21 datasheet section 17:
+  // http://www.atmel.com/images/atmel-42181-sam-d21_datasheet.pdf
+
+  int cycles;
+  uint8_t bits;
+
+  if (!_watchdog_samd_initialized)
+    _watchdog_samd_initialize_wdt();
+
+#ifdef SAMD51
+  WDT->CTRLA.reg = 0; // Disable watchdog for config
+  while (WDT->SYNCBUSY.reg)
+    ;
+#else
+  WDT->CTRL.reg = 0; // Disable watchdog for config
+  while (WDT->STATUS.bit.SYNCBUSY)
+    ;
+#endif
+
+  // You'll see some occasional conversion here compensating between
+  // milliseconds (1000 Hz) and WDT clock cycles (~1024 Hz).  The low-
+  // power oscillator used by the WDT ostensibly runs at 32,768 Hz with
+  // a 1:32 prescale, thus 1024 Hz, though probably not super precise.
+
+  if ((maxPeriodMS >= 16000) || !maxPeriodMS) {
+    cycles = 16384;
+    bits = 0xB;
+  } else {
+    cycles = (maxPeriodMS * 1024L + 500) / 1000; // ms -> WDT cycles
+    if (cycles >= 8192) {
+      cycles = 8192;
+      bits = 0xA;
+    } else if (cycles >= 4096) {
+      cycles = 4096;
+      bits = 0x9;
+    } else if (cycles >= 2048) {
+      cycles = 2048;
+      bits = 0x8;
+    } else if (cycles >= 1024) {
+      cycles = 1024;
+      bits = 0x7;
+    } else if (cycles >= 512) {
+      cycles = 512;
+      bits = 0x6;
+    } else if (cycles >= 256) {
+      cycles = 256;
+      bits = 0x5;
+    } else if (cycles >= 128) {
+      cycles = 128;
+      bits = 0x4;
+    } else if (cycles >= 64) {
+      cycles = 64;
+      bits = 0x3;
+    } else if (cycles >= 32) {
+      cycles = 32;
+      bits = 0x2;
+    } else if (cycles >= 16) {
+      cycles = 16;
+      bits = 0x1;
+    } else {
+      cycles = 8;
+      bits = 0x0;
+    }
+  }
```

Comment:
This could be rewritten as a loop that shifts `cycles` to the right to divide by two and decrements `bits`.

Reply:
Again, just copy-pasta.  If it bothers, I can refactor it, but it's simple and fast as is.